### PR TITLE
refactor(query): use InEvaluator to reuse the hashset

### LINKED ISSUE #7724: Feature: transform in scalar function to eq function
**Summary**

Description for this feature.

```
select 
   count(1) uv ,
   count(distinct t1.id) a_uv ,
   count(distinct t2.id) b_uv
from analyst.tbl6 t1
left join analyst.tbl7 t2
on t1.id = t2.id and t2.dt in ('20220809')
where t1.dt in ('20220809')
```

==>

```
select 
   count(1) uv ,
   count(distinct t1.id) a_uv ,
   count(distinct t2.id) b_uv
from analyst.tbl6 t1
left join analyst.tbl7 t2
on t1.id = t2.id and t2.dt = '20220809'
where t1.dt  = '20220809'
```

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Fixes #7724

### CODE REVIEW

**Comment by sundy-li** [comment]
The codes are a little duplicated, I'll refactor it after the new expression is added.

**Comment by RinChanNOWWW** [comment]
What's the difference between `Scalar` and `PhysicalScalar`?

File: src/query/service/src/evaluator/physical_scalar.rs [hunk 1]
```diff
@@ -40,13 +42,38 @@ impl Evaluator {
...
+                    if let EvalNode::Constant {
+                        value: DataValue::Struct(vs),
+                        ..
+                    } = &eval_args[1]
+                    {
+                        let func = if name_lower.as_str() == "not_in" {
+                            in_evalutor::create_by_values::<true>(args[0].1.clone(), vs.clone())
+                        } else {
+                            in_evalutor::create_by_values::<false>(args[0].1.clone(), vs.clone())
+                        }?;
```
Comment (RinChanNOWWW) [inline src/query/service/src/evaluator/physical_scalar.rs:62]: Why there is `args[0]` but in `Scalar::expr_scalar` is `func.arg_types[1]`?

File: src/query/service/src/evaluator/scalar.rs [hunk 2]
```diff
@@ -61,14 +63,45 @@ impl Evaluator {
...
+                            in_evalutor::create_by_values::<true>(
+                                func.arg_types[1].clone(),
+                                vs.clone(),
+                            )
+                        } else {
+                            in_evalutor::create_by_values::<false>(
+                                func.arg_types[1].clone(),
+                                vs.clone(),
+                            )
+                        }?;
```
Comment (RinChanNOWWW) [inline src/query/service/src/evaluator/scalar.rs:90]: Related to the comment above.

**Comment by RinChanNOWWW** [comment]
Such sqls will turn into error:

```
mysql> select * from t4 where b in (a, c);
ERROR 1105 (HY000): Code: 1015, displayText = Should use in_evaulator (while in processor thread 1).
mysql> select * from t4 where b+1 in (select c from t4);
ERROR 1105 (HY000): Code: 1015, displayText = Invalid child expr in subquery.

```

File: src/query/service/src/evaluator/scalar.rs [hunk 3]
```diff
@@ -61,14 +63,49 @@ impl Evaluator {
...
+
+                        return Ok(EvalNode::Function {
+                            func,
+                            args: vec![eval_args[0].clone()],
+                        });
+                    } else {
+                        return Err(ErrorCode::SyntaxException(
+                            "IN expression must have a literal array or subquery as the second argument",
+                        ));
```
Comment (RinChanNOWWW) [inline src/query/service/src/evaluator/scalar.rs:99]: We only support IN a literal array or subquery?

BTW, IN subquery seems always be converted to JOIN.

Comment (sundy-li) [reply src/query/service/src/evaluator/scalar.rs:99]: > BTW, IN subquery seems always be converted to JOIN.

Yes, so in function only covered literal array.

Comment (RinChanNOWWW) [reply src/query/service/src/evaluator/scalar.rs:99]: Will we allow non-literal array in the future? Many other DBs support it.

**Review by RinChanNOWWW — APPROVED** [review]

File: src/query/functions/src/scalars/conditionals/in_evalutor.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+use common_hashtable::HashTableKeyable;
+use common_hashtable::KeysRef;
+
+use crate::scalars::default_column_cast;
+use crate::scalars::AlwaysNullFunction;
+use crate::scalars::Function;
+use crate::scalars::FunctionContext;
+
+#[derive(Clone)]
+pub struct InEvalutorImpl<const NEGATED: bool, S: Scalar + Clone, Key: HashTableKeyable + Clone> {
```
Comment (RinChanNOWWW) [inline src/query/functions/src/scalars/conditionals/in_evalutor.rs:34]: typo: evaluator

**Review by zhang2014 — APPROVED** [review]

File: src/query/datavalues/Cargo.toml [hunk 5]
```diff
@@ -14,6 +14,7 @@ test = false
 # Workspace dependencies
 common-arrow = { path = "../../common/arrow" }
 common-exception = { path = "../../common/exception" }
+common-hashtable = { path = "../../common/hashtable" }
```
Comment (zhang2014) [inline src/query/datavalues/Cargo.toml:17]: by the way. maybe we should move `hastable` crate into `base`

Comment (sundy-li) [reply src/query/datavalues/Cargo.toml:17]: let's do it after we have saha.

**Review by RinChanNOWWW — APPROVED** [review]

**Pull request merged**